### DISCUSSION TITLE
Add some record modes that make Playtime work more like a looper

### DISCUSSION THREAD

**Discussion by helgoboss** [helgoboss/helgobox#1005] [Ideas]
Some rough ideas, unsorted, duplicated, brainstormed:

- (Destructive?) audio overdubbing as in Super8
- Retrospective recording (moving fixed-length window): Define a record length, e.g. 2 bars. Then press record and keep recording without end. When pressing stop, it would "commit" the last 2 bars. 
- NIXin: circlesongs mode
    - just start creating your clip, and press "forward" once after finishing recording loop to create another clip on top of the existing one (lossless overdub)
    - buttons for switching lanes/tracks (up/down)
    - when in an empty lane, pressing record waits until lane 1 is done, and starts recording simultaneously once lane 1 restarts playing on lane 2, if "forward" is pressed around the half duration of lane 1, it loops the block at the half-point. And vice-versa, if it's pressed at double the duration, it loops at double. Length of the loop is rounded to a subdivision of the previous track, just like in Cantor - https://youtu.be/Hvf-frOvOvI?t=492.
- Mode "Record/play/overdub" ... or rather a global "Overdub instead of recording if slot not empty"

## Old notes on this

Global record controls in app bar:
- middle or right side of app bar also contains the global record controls for triggering workflows such as:
    - Record several takes: starts recording into currently selected slot. after first recording ends (either manually or automatically depending on global settings), records into next slot TODO with same length then next. Doesn't keep playing previously recorded.
    - Record overdub into one slot: Starts recording into currently selected slot. After first recording ends, records new clip in same slot, still playing back previous one.
    - Record overdub into multiple slots: As above but records into multiple slots. Enforces non exclusive column.
- For this we need a dropdown to select the mode and a smart record button: initially, it starts the recording process. Then, when recording open end, it triggers stop of the first recording. After that, it stops recording.
- As soon as we have signal detection, we can even employ a no-button-during recording session mode that works like MIDI overdub but records into different clips. By setting a predefined length (or even by measuring a currently playing clip) and then pressing smart record: We record each cycle! whenever a cycle is empty (no input signal), we don't move on, just start recording into the same empty clip again. Related to #949

**Comment by dsorce** [comment]
Retrospective Recording would be a key decision on whether to pay for a license.

**Reply by helgoboss** [reply]
Can you describe roughly how you would expect it to work?

**Reply by dsorce** [reply]
If I could have definable bar length commit buttons. 
1 bar 2 bars 4 bars 8 bars... 16, 32
This software is dead, so  would be a replacement. 
Play as long as you want, commit when you hear the the loop you like. 
https://youtube.com/clip/Ugkxt_AW6in9U8HqNTYSVMy_NMzpGByneV60?si=c3Os8U-9vxLd0Iku

**Reply by helgoboss** [reply]
Okay, I just made a small concept for retrospective recording. Check it out whether that would work for you. If not, write a user scenario (such as the one below) that describes what you want.

Scenario with start timing "Quantized: 1 bar":

1. User defines a record length, e.g. 2 bars (we already have a record length option)
2. User switches record mode to "Retrospective recording"
3. User triggers recording (either via slot record button or smart record button)
4. Playtime starts recording bar 1 as soon as the next bar is reached
5. Playtime records bar 2
6. Playtime records bar 3
7. Playtime records bar 4. During that time, user stops recording.
8. At the end of bar 4, Playtime keeps bar 3 and 4 and immediately starts playing bar 3.


Scenario with start timing "Immediate":

- Doesn't make sense, I think. Probably should be prevented.

**Reply by dsorce** [reply]
Scenario with start timing "Quantized: 1 bar":
This sounds great.  Would love to try this out.

**Reply by RibbleFripFnibbulator** [reply]
+1 for audio overdubbing and record modes. That would be amazing. Ideally a Replace with the option to immediate replace playback or to hearing playback during the replace, i.e. Substitute function from Mobius looper. Things like Insert, Append would also be amazing.

**Comment by RibbleFripFnibbulator** [comment]
Immediate Next-Loop Overdub/Record:
1. Silence mode
2. Record into a blank slot (exclusive mode disabled)
3. ????
4. Tempo is set, first clip plays back, and the clip below it is already recording at the same tempo immediately on FIRST playback

So far with all my methods I am not able to achieve this. I put together a Trigger->Browse down->Trigger script hoping that maybe if I send the commands quickly enough in succession, the second loop trigger's lateness would be forgiven and it would still enter record mode immediately. However it appears this is not possible. Playtime will wait until one "cycle" passes before the second loop begins to record.

Can this be implemented? An immediate "next layer overdub" if you will? This would allow Playtime to be used for faster live-looping.

From https://github.com/helgoboss/helgobox/discussions/1281

**Comment by xlq** [comment]
Clip launchers and loopers are fundamentally at odds with how they're used, but there's a way to reconcile that.

At the end of recording a loop, a looper can record any note releases or reverb tails by overdubbing them over the start of the loop, so that the loop is seamless and there's no abrupt cut-off. Clip launchers don't tend to do that but instead cut off abruptly - but why? Because clip launchers are intended to bring parts in and out. If your loop has tails at the beginning, stopping it will cut the tails off anyway, and worse, when you start it, the tails will suddenly sound as soon as the loop comes back in. Slowly fading parts in and out is one solution to that, but dictates a certain style.

In Reaper, if you make a 4 bar time selection, enable loop playback and hit record, you can record loops with overdubs just like a looper pedal, except that it's all recorded to one audio file, and multiple items are placed on the timeline so that the one audio file is wrapped around the loop. I like this because it's immediately usable as a loop, but you can also unroll the loop along the timeline so that you can start playing it without the tail, loop it 1 or more times, and end with just the tail.

Unfortunately you can't use Reaper as a proper looper like this because the audio doesn't start playing back until you stop recording. There's an option to create multiple audio files - one for each loop - but there's still a delay adding each loop to the timeline and it seems to chop tiny bits off the end of the loop. But it makes me think Playtime can work similarly.

There are two uses for overdubbing, and it seems useful to distinguish between them: the first is to record tails, upbeats etc. The second is to add more layers to the same loop. This second use is perhaps better served by adding new stacked clips, while the first might be better served with an extension to Playtime's upbeat feature.

Consider that currently, an audio clip in Playtime has an audio file and a start and end point. If a clip instead had four points - start, loop start, loop end and end - then playback could proceed as follows:

1. Playback starts at the start point at a time such that the loop start point coincides with the launch time (this allows upbeats or just a note that's hit slightly too early).
2. Playback proceeds until the loop end point. At the loop end point, if the clip is queued to stop, playback proceeds to the end point and then stops. If the clip is not queued to stop, playback still continues to the end point but *additionally* a second playhead starts from the loop start point.
3. (The second playhead should actually start at the start point before the first playhead reaches the loop end point, to include the upbeat).

When exporting to the timeline, the loops can also be arranged like this, to give perfect loops with tails in the right places and no abrupt cutoffs or rogue tails.

Unfortunately I don't know how feasible it is to do this in Playtime because it would require playing back an audio file that's still being written.

**Comment by paunoliini** [comment]
Is there any guess when the destructive audio overdubbing is going to be available? That's the only thing I'm waiting cause even Ableton can't do this. The other one is editing functions in clips. That would be so amazing.

**Reply by helgoboss** [reply]
My guess: Not before end of year. Sorry, I'm currently in the phase of funding the next development cycle and that takes a while.

**Comment by paunoliini** [comment]
I hope you find a funding! Good luck.

- Jussi

**Reply by helgoboss** [reply]
It can't really be found, unfortunately. What I mean is to earn some money by working on more profitable things than Playtime for a while ;)

If you want to help accelerate things, the best way is to tell other users about it. That way, Playtime gets more profitable and I can work on it more frequently.

**Comment by helgoboss** [comment]
Also see #1436.

**Comment by fiaganja** [comment]
I want to add that a way to either control what is recorded (the output input, or input fx) of a track, or simply just mirror what is chosen for that track would be so good. It is very limited that we can only record the input and super super time consuming to create multiple tracks every time you want to record the output of an instrument or sampler into a slot :/

**Comment by larryseyer** [comment]
Request to give Playtime 2 ability to emulate Quantiloop Pro's serial looping mode where:

1. Initial clip recording defines scene and clip length.

2. All other clips for that scene respect clip 1's length for scene switching.

3. Subsequent recording to empty clips after first initial recorded clip will loop within current scene but respect scene length for scene switch.

4. Subsequent recording to a previously recorded clip after initial clip recording is always in overdub mode and can happen at any point within the clip.

5. Scene switching occurs at scene boundaries but can be overridden by a 2nd request for a scene switch.  This can happen immediately, or on measure, beat boundaries.

6. Stop and Start respect last selected scene and clip.

(edited for clarity and spelling errors)
